scripts/install-cni.sh

```diff
@@ -131,19 +131,119 @@ else
+  local ip=$1
+
+  if [[ "${ip:-}" =~ ${IPV6_VALIDATION_REGEXP} ]] ; then
+    return 0
+  else
+    return 1
+  fi
+}
+
+#######################################
+# Replaces `@subnets` and `@routes` placeholders using `.spec.podCIDRs` from node json.
+# Arguments:
+#   cni spec template
+#   node json from kube-apiserver
+# Outputs:
+#   template with substituted placeholders
+#######################################
+function fillSubnetsInCniSpec {
+  local cni_spec=$1
+  local node=$2
+  local node_ipv6_addr=$3
+
+  local SUBNETS_REPLACEMENT=""
+  local ROUTES_REPLACEMENT=""
+
+  for subnet in $(jq -r '.spec.podCIDRs[]' <<<"$node") ; do
+    if is_ipv4_range "$subnet" ; then
+      if [ "" != "${SUBNETS_REPLACEMENT}" ] ; then
+        SUBNETS_REPLACEMENT="${SUBNETS_REPLACEMENT}, "
+        ROUTES_REPLACEMENT="${ROUTES_REPLACEMENT}, "
+      fi
+      SUBNETS_REPLACEMENT="${SUBNETS_REPLACEMENT}[{\"subnet\": \"${subnet}\"}]"
+      ROUTES_REPLACEMENT="${ROUTES_REPLACEMENT}{\"dst\": \"0.0.0.0/0\"}"
+    elif is_ipv6_range "$subnet" ; then
+      if [ "" != "${SUBNETS_REPLACEMENT}" ] ; then
+        SUBNETS_REPLACEMENT="${SUBNETS_REPLACEMENT}, "
+        ROUTES_REPLACEMENT="${ROUTES_REPLACEMENT}, "
+      fi
+      SUBNETS_REPLACEMENT="${SUBNETS_REPLACEMENT}[{\"subnet\": \"${subnet}\"}]"
+      ROUTES_REPLACEMENT="${ROUTES_REPLACEMENT}{\"dst\": \"::/0\"}"
+    else
+      echo "[ERROR] Subnet $subnet is not a valid IP range"
+      exit 1
+    fi
+  done
+
+  # Directpath usecase
+  if [ "$ENABLE_IPV6" == "true" ] ; then
+    # Directpath adds IPv6 subnet and route derived from host with fixed range
+    # of /112 even when it is not specified in node's .spec.podCIDRs
+    if [ -n "${node_ipv6_addr:-}" ] && [ "${node_ipv6_addr}" != "null" ]; then
+      SUBNETS_REPLACEMENT="${SUBNETS_REPLACEMENT}, [{\"subnet\": \"${node_ipv6_addr}/112\"}]"
+      ROUTES_REPLACEMENT="${ROUTES_REPLACEMENT}, ${CNI_SPEC_IPV6_ROUTE:-{\"dst\": \"::/0\"\}}"
+    fi
+  fi
+
+  cni_spec=${cni_spec//@subnets/$SUBNETS_REPLACEMENT}
+  cni_spec=${cni_spec//@routes/$ROUTES_REPLACEMENT}
+
+  echo "${cni_spec}"
+}
+
+STACK_TYPE=$(jq '.metadata.labels."cloud.google.com/gke-stack-type"' <<<"$response")
+echo "Node stack type label: '${STACK_TYPE:-}'"
 
-if [[ "${ipv4_subnet:-}" =~ ^\"[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*\.[0-9][0-9]*/[0-9][0-9]*\"$ ]]; then
+node_ipv6_addr=''
+if [ "$ENABLE_IPV6" == "true" ] || [ "${STACK_TYPE:-}" == '"IPV4_IPV6"' ]; then
+  node_ipv6_addr=$(curl -s -k --fail "http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/?recursive=true" -H "Metadata-Flavor: Google" | jq -r '.ipv6s[0]' ) ||:
+fi
+
+cni_spec=$(fillSubnetsInCniSpec "$cni_spec" "$response" "$node_ipv6_addr")
+ipv4_subnet=$(jq -r '.spec.podCIDR' <<<"$response")
```

Comment:
in an ipv6 only cluster this will be an ipv6 subnet

Reply:
Can we unify this logic together with `fillSubnetsInCniSpec`?

Reply:
I extracted all logic for old IPAM replacements to function `fillSubnetsInCniSpecLegacyTemplate`. I would rather not mix the old and new template logic together.